### PR TITLE
158-simplify-cloudwatch-to-graphite-app-remove-leadbutt-dependency

### PR DESCRIPTION
**Pull request by benvand**
Refactor/ simplification of the app:

* Remove `pipe-watchdog.py` entirely
  * We're no longer using pipes to direct the output. An error should be output to the cloudwatch error logs on failure `cf logs --recent digitalmarketplace-cloudwatch-to-graphite-preview`
  * Remove this from the run command in `manifest.yml.example`

* Add `flake8` dependency and `.flake8` file in line with GDS recommendations
* Activate travis service on repo to check Flake8 rules and requirements files
  * Add `.travis.yml`

* Add `Makefile` with shortcuts for running requirements freeze and flake8
* Update `README.md`
* Change requirements handling
  * Add `requirements-app.txt`
  * Ensure `requirements.txt` contains full pinned requirements
  * Add `flake8` requirement for devs

* Rewrite `app.py`
  * Add function descriptions to all functions
  * Remove dependency on `leadbutt`, `signal` and `subprocess`
  * Use `boto3` instead of `leadbutt`s `boto` (https://github.com/crccheck/cloudwatch-to-graphite/blob/master/leadbutt.py#L25)
  * Remove `get_timestamp` method. The same can be achieved with `time.time`
  * `logger.warn` is deprecated, use `logger.warning` instead
  * Bring string formatting in line with rest of strings in file
  * Rename `initialize_metrics ` to `create_hostedgraphite_base_metrics` for better description
  * Remove `call_leadbutt` function and replace with code calling aws ourselves. Extrapolated from (https://github.com/crccheck/cloudwatch-to-graphite/blob/master/leadbutt.py)
  *  Change `loglevel` to `INFO` this is all that is required now we're not piping output
  * Move retry decorator on to wrapper method in `if __name__` block

* Update `config.yaml.j2` template
  * `Dimensions` default = `{}` to avoid type manipulations
  * Remove `Auth` section from config. This can be hard coded until we have any need to specify it as a config variable
  * Remove `Options.Count` again, this can be hard coded. It's only used in one place and we've never needed to change it
* Replace `retrying` with `tenacity`
  * `retrying` is stale. `tenacity` is a community maintained fork. (https://github.com/rholder/retrying/issues/79)

## TL:DR;

Extrapolate code out of leadbutt
This code drops leadbutts retry functionality. Boto3 has the capacity to retry failed attempts. We should prefer this.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,7 +3,8 @@ Digital Marketplace Cloudwatch to Graphite
 
 ## Purpose
 
-Ships Cloudwatch metrics to Hosted Graphite using https://github.com/crccheck/cloudwatch-to-graphite.
+Ships Cloudwatch metrics to Hosted Graphite.
+Heavily draws on the work of https://github.com/crccheck/cloudwatch-to-graphite
```
Comment (katstevens) [inline README.md]: Can we be more specific and say which bits we have been 'inspired by'?

File: app.py [hunk 2]
```diff
@@ -13,58 +12,113 @@
...
+        config_metric_entry['Statistics'].lower()
+    )
+    hostedgraphite_base_metric = "{} 0.0 {}".format(hostedgraphite_metric_name, timestamp)
+    return hostedgraphite_base_metric
 
 
-@retry(wait_fixed=60000, retry_on_result=lambda res: res is None)
-def call_leadbutt():
+def create_hostedgraphite_base_metrics(config):
+    """Take a metric entry from the config format it into a base metric used to initialise the metric on hostedgraphite.
```
Comment (katstevens) [inline app.py:49]: Could we simplify this to `"""Initialise each base metric in the config with a timestamp"""`?

File: app.py [hunk 3]
```diff
@@ -13,58 +12,113 @@
...
+    ).lower()
+    return '{0} {1} {2}'.format(
+        hostedgraphite_metric_name,
+        result[cloudwatch_metric_datapoint['Statistics']],
+        result['Timestamp'].strftime('%s')
+    )
+
+
+def get_metric_from_cloudwatch(client, metric):
+    """Call the client once for th supplied metric."""
```
Comment (katstevens) [inline app.py:75]: Typo `th`?

File: tests/test_app.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+from .fixtures import config
+
+
+class TestApp:
+
+    def setup_method(self):
+        # Mock out boto3 cloudwatch client
+        self.client = boto3.client('cloudwatch', region_name="eu-west-1")
+        stubber = Stubber(self.client)
```
Comment (katstevens) [inline tests/test_app.py]: Interesting! Is this a better option than `moto`?

File: tests/test_app.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        stubber.activate()
+
+        # Mock out send_to_hostedgraphite method
+        self.send_to_hostedgraphite_mock = mock.patch('app.send_to_hostedgraphite')
+        self.send_to_hostedgraphite_mock.start()
```
Comment (katstevens) [inline tests/test_app.py:35]: Should we have an equivalent `.stop()` in a `teardown_method()`?

**Review by katstevens — COMMENTED** [review]
Nice work on this - just a few minor tweaks suggested.

File: app.py (hunk 2 above)
Comment (benvand) [reply app.py:49]: Almost, but we're also initialising it with a base number.

They come out like:
`cloudwatch.request_time_buckets.preview.antivirus-api.request_time_bucket_0.sum 0.0 1537113600`
`{metric_name} {base_value} {timestamp}`

Setting a starting point for the metric.

File: tests/test_app.py (hunk 4 above)
Comment (benvand) [reply tests/test_app.py]: I think so in so much as it prevents another dependency and should be supported for the lifetime of the `boto3` lib. Pretty nifty, but it hasn't been around for that long. I expect the dependency in utils was added before this existed.

File: README.md (hunk 1 above)
Comment (benvand) [reply README.md]: Done 👍

File: tests/test_app.py (hunk 5 above)
Comment (benvand) [reply tests/test_app.py:35]: Good shout 👍 Done

File: app.py (hunk 2 above)
Comment (katstevens) [reply app.py:49]: OK - `Initialise each base metric in the config with a timestamp and base number` then? 😸 The current docstring is a bit hard to read.

File: README.md [hunk 6]
```diff
@@ -3,9 +3,10 @@ Digital Marketplace Cloudwatch to Graphite
 
 ## Purpose
 
-Ships Cloudwatch metrics to Hosted Graphite using https://github.com/crccheck/cloudwatch-to-graphite.
+Ships Cloudwatch metrics to Hosted Graphite.
+Heavily on the work of https://github.com/crccheck/cloudwatch-to-graphite and the `leadbutt` script. Particularly config parsing to define metrics.
```
Comment (katstevens) [inline README.md]: 👍 (Though I think it's now missing a word: `Heavily based`?)

**Review by katstevens — APPROVED** [review]
👍

**Pull request merged**